scan-external-contacts/nodejs/index.js

```diff
@@ -0,0 +1,325 @@
+const gcloudContactsScan = async (cursor) => {
+
+    // Defining the default values for the result structure
+    let result = {
+        error: false,
+        errorMsg: "",
+        entities: [],
+        isEmpty: true,
+        isFinished: true,
+        nextCursor: null,
+        startTime: 0
+    };
+
+    try {
+        result.startTime = Date.now();
+
+        let opts = {
+            'limit': maxRecordsPerPage
+        };
+        // No opts.cursor on first scan - cursor == null
+        if (cursor) {
+            opts.cursor = cursor;
+        }
+
+        // External Contacts Scan API request
+        let scanData = await contactsApi.getExternalcontactsScanContacts(opts);
+        // If the request is successful, we analyze the response
+
+        // The response may contain an empty "entities" array in certain conditions (congestion on database, contacts deleted before the equest was performed, ...)
+        if (scanData.entities && scanData.entities.length > 0) {
+            result.isEmpty = false;
+            result.entities = scanData.entities;
+        }
+
+        // If the response contains an "after" attribute inside "cursors" (cursors.after), then, there are still some External Contacts (data) to retrieve
+        // When the response does not contain an "after" attribute inside "cursors", it means it is the last page of scan
+        if (scanData.hasOwnProperty("cursors") && scanData.cursors.hasOwnProperty("after")) {
+            result.isFinished = false;
+            result.nextCursor = scanData.cursors.after;
+        }
+
+        // Reset of some global vars - counters for retry
+        consecutiveLoginRetryAttempts = 0;
+        consecutiveRateLimitRetryAttempts = 0;
+
+        return result;
+    }
+    catch (error) {
+        console.log('External Contacts Scan error - Error:', JSON.stringify(error, null, 2));
+
+        if (error.status && error.status === 401) {
+            // Login - Invalid token, Token expired, ...
+            // Exit Application if too many consecutive login failures
+            if (consecutiveLoginRetryAttempts >= maxLoginRetryAttempts) {
+                result.errorMsg = "Max consecutive login failures. Exiting...";
+                result.error = true;
+                return result;
+            } else {
+                consecutiveLoginRetryAttempts++;
+
+                // Trying to login again
+                let isLoggedIn = await login();
+
+                if (isLoggedIn) {
+                    // Try to request the same data again (current value of cursor)
+                    let retryScanData = await gcloudContactsScan(cursor);
+                    return retryScanData;
+                } else {
+                    result.errorMsg = "Unsuccessful Relogin attempt. Exiting...";
+                    result.error = true;
+                    return result;
+                }
+            }
+        } else if (error.status && error.status === 429) {
```

Comment:
The JS SDK will have built-in rate limit retrying when DEVENGAGE-271 is implemented. Preferably, remove all of the logic related to this. If this tutorial can't function without handling rate limiting, please implement rate limiting logic per the documented best practices: https://developer.mypurecloud.com/api/rest/rate_limits.html#best_practices_for_retrying_requests

Reply:
Nice.